src/main/java/com/aws/greengrass/builtin/services/pubsub/SubscriptionTrie.java

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.aws.greengrass.builtin.services.pubsub;
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+
+import java.util.Collections;
+import java.util.LinkedHashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Trie to manage subscriptions.
+ */
+public class SubscriptionTrie {
+    private static final char TOPIC_LEVEL_SEPARATOR = '/';
+    private static final char SINGLE_LEVEL_WILDCARD = '+';
+    private static final char MULTI_LEVEL_WILDCARD = '#';
+
+    private final Map<String, SubscriptionTrie> children = new ConcurrentHashMap<>();
+    @SuppressWarnings("PMD.UnusedPrivateField")
+    @SuppressFBWarnings("URF_UNREAD_FIELD")
+    private final String value;
+    private final Set<Object> callbacks;
+
+    /**
+     * Construct.
+     */
+    public SubscriptionTrie() {
+        this("");
+    }
+
+    private SubscriptionTrie(String value) {
+        this.value = value;
+        this.callbacks = ConcurrentHashMap.newKeySet();
+    }
+
+    private SubscriptionTrie lookup(String topic) {
+        SubscriptionTrie current = this;
+        for (String topicLevel : topic.split(String.valueOf(TOPIC_LEVEL_SEPARATOR))) {
+            current = current.children.get(topicLevel);
+            if (current == null) {
+                return null;
+            }
+        }
+        return current;
+    }
+
+    /**
+     * Check if trie contains a topic.
+     *
+     * @param topic topic
+     * @return if trie contains topic
+     */
+    public boolean containsKey(String topic) {
+        return lookup(topic) != null;
+    }
+
+    /**
+     * Remove entry for one callback.
+     *
+     * @param topic topic
+     * @param cb    callback
+     * @return if changed after removal
+     */
+    public boolean remove(String topic, Object cb) {
+        return remove(topic, Collections.singleton(cb));
+    }
+
+    /**
+     * Remove entry for a set of callbacks.
+     *
+     * @param topic topic
+     * @param cbs   callbacks
+     * @return if changed after removal
+     */
+    public boolean remove(String topic, Set<Object> cbs) {
+        SubscriptionTrie sub = lookup(topic);
+        if (sub == null) {
+            return false;
+        }
+        return sub.callbacks.removeAll(cbs);
+    }
+
+    /**
+     * Size of trie.
+     *
+     * @return size
+     */
+    public int size() {
+        int[] size = {this.callbacks.size()};
+        children.forEach((s, t) -> {
+            size[0] += t.size();
+        });
+        return size[0];
+    }
+
+    /**
+     * Add a topic callback.
+     *
+     * @param topic topic
+     * @param cb    callback
+     * @return true
+     */
+    public boolean add(String topic, Object cb) {
```

Comment:
For `remove`, we have overloaded method for a single or a set of `cb`. I think it's a good idea to be consistent and overload `put` instead of calling this `add`